File: SubSets_Backtracking.java
//Time Complexity: O(2^N)
//Space Complexity: O(N)

import java.util.*;

public class SubSets_Backtracking {
	class Solution {
		List<List<Integer>> result;

		public List<List<Integer>> subsets(int[] nums) {
			result = new ArrayList<>();
			if (nums == null || nums.length == 0)
				return result;
			backtrack(nums, new ArrayList<>(), 0);
			return result;
		}

		private void backtrack(int[] nums, List<Integer> temp, int start) {
			// base
			// logic
			result.add(new ArrayList<>(temp));
			for (int i = start; i < nums.length; i++) {
				// action
				temp.add(nums[i]);
				// recurse
				backtrack(nums, temp, i + 1);
				// backtrack
				temp.remove(temp.size() - 1);
			}
		}
	}
}

File: Subsets_backtracking.py
# TC: O(2 ^ N) since we are exploring 2 options for every number in the input list. 
# SC: O(N) where N is the space used by the path list. In worst case, it will be storing all input numbers, hence O(N).

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        if not nums or len(nums) == 0:
            return 
        
        self.result = []
        
        def helper(index, path):
            if index == len(nums):
                self.result.append(path.copy())
                return
            
#             dont choose
            helper(index + 1, path)
#               choose
#           action
            path.append(nums[index])
#           recurse
            helper(index + 1, path)
#           backtrack
            path.pop()
    
        helper(0, [])
        return self.result
